uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/TOILRequestTest.php

```diff
@@ -0,0 +1,150 @@
+      'max_leave_accrual' => 10
+    ]);
+
+    $this->leaveBalanceChangeService = new LeaveBalanceChangeService();
+
+    $this->leaveContact = 1;
+    $this->registerCurrentLoggedInContactInSession($this->leaveContact);
+    CRM_Core_Config::singleton()->userPermissionClass->permissions = [];
+  }
+
+  public function testCreateAlsoCreatesTheLeaveRequestAndTheTOILRequestBalanceChange() {
+    $toilRequestService = $this->getTOILRequestService();
+    $params = $this->getDefaultParams();
+    $toilRequest = $toilRequestService->create($params, false);
+
+    LeaveRequest::findById($toilRequest->leave_request_id);
+    $toilBalanceChange = $this->findToilRequestBalanceChange($toilRequest->id);
+    $this->assertEquals($toilBalanceChange->amount, $params['toil_to_accrue']);
+  }
+
+  public function testDeleteAlsoDeletesTheLeaveRequestItsBalanceChangesAndDates() {
+    $params = $this->getDefaultParams([
+      'from_date' => CRM_Utils_Date::processDate('+1 day'),
+      'to_date' => CRM_Utils_Date::processDate('+3 days'),
+    ]);
+    $toilRequest = TOILRequestFabricator::fabricateWithoutValidation($params, true);
+
+    $leaveRequest = LeaveRequest::findById($toilRequest->leave_request_id);
+    $toilBalanceChange = $this->findToilRequestBalanceChange($toilRequest->id);
+    $leaveRequestDates = $leaveRequest->getDates();
+    $this->assertCount(3, $leaveRequestDates);
+    $this->assertInstanceOf(LeaveBalanceChange::class, $toilBalanceChange);
+
+    //delete the TOIL
+    $service = $this->getTOILRequestService();
+    $service->delete($toilRequest->id);
+
+    $this->assertNull($this->findToilRequestBalanceChange($toilRequest->id));
+    $dates = $leaveRequest->getDates();
+    $this->assertCount(0, $dates);
+
+    try {
+      LeaveRequest::findById($leaveRequest->id);
+      $this->fail("Expected to not find the LeaveRequest with {$leaveRequest->id}, but it was found");
+    } catch (Exception $e) {
+      return;
+    }
+
+    try {
+      TOILRequest::findById($toilRequest->id);
+      $this->fail("Expected to not find the TOILRequest with {$toilRequest->id}, but it was found");
+    } catch (Exception $e) {
+      return;
+    }
+  }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You are not allowed to change the type of a request
+   */
+  public function testCreateThrowsAnExceptionWhenLeaveApproverUpdatesAbsenceTypeForTOILRequest() {
+    $contactID = 5;
+    $params = $this->getDefaultParams(['contact_id' => $contactID]);
+    $toilRequest = TOILRequestFabricator::fabricateWithoutValidation($params, true);
+
+    $absenceType = AbsenceTypeFabricator::fabricate([
+      'allow_accruals_request' => true,
+      'max_leave_accrual' => 10
+    ]);
+
+    $params['id'] = $toilRequest->id;
+    $params['type_id'] = $absenceType->id;
+
+    $toilRequestService = $this->getTOILRequestServiceWhenCurrentUserIsLeaveManager();
+    $toilRequestService->create($params, false);
+  }
```

Comment:
I'm not sure if this method is really necessary, since the behavior is the same as the LeaveRequest Service, so we're just duplicating the test from that service